dlib compile error with CUDA 10.2 (identifier "dlib::sqrt_2" is undefined in device code)

<!-- ================================================================== -->
<!-- ==================  BUG/PROBLEM REPORT TEMPLATE ================== -->
<!-- ================================================================== -->


<!-- Delete the above instructions and then provide a general summary of the issue in the Title above.  Then fill out the template below.  FAILURE TO FILL OUT EVERY PART WILL RESULT IN YOUR ISSUE BEING CLOSED -->

## Expected Behavior
<!--- Tell us what should happen.  What were you doing?  What part of dlib are you using?  What do you think should happen? -->
dlib should compile with CUDA

## Current Behavior
<!--- Tell us what happens instead of the expected behavior.  If you get an error, include the entire error message in the bug report.  DO NOT POST SCREEN SHOTS.  Paste in the text instead.   If the issue is some kind of build problem, include the entire CMake output along with the error message. -->
It does not compile
```
running install
running bdist_egg
running egg_info
writing tools/python\dlib.egg-info\PKG-INFO
writing dependency_links to tools/python\dlib.egg-info\dependency_links.txt
writing top-level names to tools/python\dlib.egg-info\top_level.txt
package init file 'tools\python\dlib\__init__.py' not found (or not a regular file)
reading manifest file 'tools/python\dlib.egg-info\SOURCES.txt'
reading manifest template 'MANIFEST.in'
no previously-included directories found matching 'tools\python\build*'
writing manifest file 'tools/python\dlib.egg-info\SOURCES.txt'
installing library code to build\bdist.win-amd64\egg
running install_lib
running build_py
running build_ext
Building extension for Python 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
Invoking CMake setup: 'cmake C:\Users\Vayun Mathur\OneDrive\Documents\code\python\dlib\tools\python -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\Users\Vayun Mathur\OneDrive\Documents\code\python\dlib\build\lib.win-amd64-3.8 -DPYTHON_EXECUTABLE=C:\Users\Vayun Mathur\anaconda3\envs\AI\python.exe -DCMAKE_CXX_FLAGS=-DCUDA_HOST_COMPILER=/usr/bin/gcc-6 -DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\Users\Vayun Mathur\OneDrive\Documents\code\python\dlib\build\lib.win-amd64-3.8 -A x64'      
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19041.
-- The C compiler identification is MSVC 19.27.29111.0
-- The CXX compiler identification is MSVC 19.27.29111.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: C:/Users/Vayun Mathur/anaconda3/envs/AI/python.exe (found version "3.8.5") 
-- Found PythonLibs: C:/Users/Vayun Mathur/anaconda3/envs/AI/libs/Python38.lib
-- pybind11 v2.2.2
-- Using CMake version: 3.18.2
-- Compiling dlib version: 19.21.99
-- SSE4 instructions can be executed by the host processor.
-- AVX instructions can be executed by the host processor.
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Enabling AVX instructions
-- Searching for BLAS and LAPACK
-- Searching for BLAS and LAPACK
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE  
-- Found CUDA: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.2 (found suitable version "10.2", minimum required is "7.5") 
-- Looking for cuDNN install...
-- Found cuDNN: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.2/lib/x64/cudnn.lib
-- Building a CUDA test project to see if your compiler is compatible with CUDA...
-- Building a cuDNN test project to check if you have the right version of cuDNN installed...
-- Enabling CUDA support for dlib.  DLIB WILL USE CUDA
-- C++11 activated.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/Vayun Mathur/OneDrive/Documents/code/python/dlib/build/temp.win-amd64-3.8/Release
Invoking CMake build: 'cmake --build . --config Release -- /m'
Microsoft (R) Build Engine version 16.7.0+b89cb5fde for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building NVCC (Device) object dlib_build/CMakeFiles/dlib.dir/cuda/Release/dlib_generated_cuda_dlib.cu.obj
  cuda_dlib.cu
C:/Users/Vayun Mathur/OneDrive/Documents/code/python/dlib/dlib/cuda/cuda_dlib.cu(1504): error : identifier "dlib::pi" is undefined in device code [C:\Users\Vayun Mathur\OneDrive\Documents\code\python\dlib\build\temp.win-amd64-3.8\Release
\dlib_build\dlib.vcxproj]
  
C:/Users/Vayun Mathur/OneDrive/Documents/code/python/dlib/dlib/cuda/cuda_dlib.cu(1504): error : identifier "dlib::sqrt_2" is undefined in device code [C:\Users\Vayun Mathur\OneDrive\Documents\code\python\dlib\build\temp.win-amd64-3.8\Rel 
ease\dlib_build\dlib.vcxproj]
  
  2 errors detected in the compilation of "C:/Users/VAYUNM~1/AppData/Local/Temp/tmpxft_00005f64_00000000-10_cuda_dlib.cpp1.ii".
  cuda_dlib.cu
  CMake Error at dlib_generated_cuda_dlib.cu.obj.Release.cmake:280 (message):
    Error generating file C:/Users/Vayun
    Mathur/OneDrive/Documents/code/python/dlib/build/temp.win-amd64-3.8/Release/dlib_build/CMakeFiles/dlib.dir/cuda/Release/dlib_generated_cuda_dlib.cu.obj


Traceback (most recent call last):
  File "setup.py", line 223, in <module>
    setup(
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\site-packages\setuptools\__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\site-packages\setuptools\command\install.py", line 67, in run
    self.do_egg_install()
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\site-packages\setuptools\command\install.py", line 109, in do_egg_install
    self.run_command('bdist_egg')
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\site-packages\setuptools\command\bdist_egg.py", line 167, in run
    cmd = self.call_command('install_lib', warn_dir=0)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\site-packages\setuptools\command\bdist_egg.py", line 153, in call_command
    self.run_command(cmdname)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\site-packages\setuptools\command\install_lib.py", line 11, in run
    self.build()
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\command\install_lib.py", line 107, in build
    self.run_command('build_ext')
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 135, in run
    self.build_extension(ext)
  File "setup.py", line 175, in build_extension
    subprocess.check_call(cmake_build, cwd=build_folder)
  File "C:\Users\Vayun Mathur\anaconda3\envs\AI\lib\subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--', '/m']' returned non-zero exit status 1.
```

## Steps to Reproduce
<!--- Provide an unambiguous set of steps to reproduce this problem.  What exactly do I have to type to reproduce your error?  If you don't post complete instructions you are much less likely to get help. -->

1. Install CUDA 10.2
2. Install cuDNN 7.6.5
3. `git clone https://github.com/davisking/dlib.git`
4. `python setup.py install`


* **Version**: <!-- What version of dlib? --> 19.21.99
* **Where did you get dlib**: <!-- Did you get it from an official source like dlib.net, this github repo, or somewhere else? --> This github repo
* **Platform**: <!-- include something like `lsb_release -a` output, or if Windows, version and 32-bit or
  64-bit --> Windows 10 64 bit 10.0.19041 Build 19041
* **Compiler**: <!-- What compiler are you using?  What version of that compiler? --> CMake 3.18.2

--- a/dlib/cuda/cpu_dlib.cpp
+++ b/dlib/cuda/cpu_dlib.cpp
@@ -1711,7 +1711,7 @@ namespace dlib
             const tensor& gradient_input
         )
         {
-            const float beta = 1.0f / std::sqrt(pi) / sqrt_2;
+            const float beta = 1.0f / std::sqrt(2.0f * pi);
             const auto compute_gradient = [beta](float x)
             {
                 const float cdf = 0.5f*(1.0f + std::erf(x/sqrt_2));

--- a/dlib/cuda/cuda_dlib.cu
+++ b/dlib/cuda/cuda_dlib.cu
@@ -1501,7 +1501,7 @@ namespace dlib
 
         __device__ float gelu_compute_gradient(float x)
         {
-                const float beta = 1.0f / std::sqrt(pi) / sqrt_2;
+                const float beta = 1.0f / std::sqrt(2.0f * pi);
                 const float cdf = normcdf(x);
                 const float pdf = beta*std::exp(-0.5f*x*x);
                 return cdf + x * pdf;

--- a/dlib/cuda/cuda_dlib.cu
+++ b/dlib/cuda/cuda_dlib.cu
@@ -3,8 +3,8 @@
 
 #include "cuda_utils.h"
 #include "cuda_dlib.h"
-#include <math_constants.h>
 #include "cudnn_dlibapi.h"
+#include <math_constants.h>
 
 
 namespace dlib 

--- a/dlib/cuda/cuda_dlib.cu
+++ b/dlib/cuda/cuda_dlib.cu
@@ -3,6 +3,7 @@
 
 #include "cuda_utils.h"
 #include "cuda_dlib.h"
+#include <math_constants.h>
 #include "cudnn_dlibapi.h"
 
 
@@ -1501,7 +1502,7 @@ namespace dlib
 
         __device__ float gelu_compute_gradient(float x)
         {
-                const float beta = 1.0f / std::sqrt(2.0f * pi);
+                const float beta = 1.0f / CUDART_SQRT_2PI;
                 const float cdf = normcdf(x);
                 const float pdf = beta*std::exp(-0.5f*x*x);
                 return cdf + x * pdf;

